### ISSUE TITLE
Possible reasons for loss function not going down

### ISSUE THREAD

**Issue by sarimmehdi** [TengdaHan/DPC#7]
Hello. I am using your network for self-supervision representation on the kitti dataset. Even after 22 epochs the loss function and top1 accuracy barely change. What possible reasons could explain this?

**Comment by TengdaHan** [comment]
It's hard to answer this question. But a few checks you can do:
1. do you have enough batch negative pairs? e.g. I use batch size >16 per gpu.
2. is your problem setting too hard? e.g. how much is the final feature size (`TxHxWx256` or similar), what size of motion do you expect the model to capture (tiny object motion? if so, probably try contrastive loss on larger feature map)
3. print some of the similarity scores during training https://github.com/TengdaHan/DPC/blob/7bb73f4a934bbf34eda10a6102629bacb9b4e348/dpc/model_3d.py#L83 the tensor with size `(B, T, S**2, B, T, S**2)` before and after softmax. The first 3 dimensions are for prediction, the last 3 dimensions are for 'ground-truth' feature extracted from future data. During training you should observe:
    * First, similarity score for batch negatives (`B!=B`) quickly becomes lower
    * Second, similarity score for spatial negatives (`B==B`, but `S**2!=S**2`) becomes lower
    * Third, model trying to distinguish temporal negatives (`B==B`, `S**2==S**2`, but `T!=T`)

**Comment by TengdaHan** [comment]
What downstream tasks are you trying to do on Kitti?

**Comment by sarimmehdi** [comment]
My idea is to extract features from the first N frames and then concatenate those features (after pooling and fully connected layer) with the corresponding bounding box encoded features (these are obviously extracted using a traditional encoder RNN framework and not yours) before sending them at each time step of a decoder to predict bounding box coordinates

**Comment by sarimmehdi** [comment]
- I use batch size of 32. 
- The final feature size is 5 by 4 by 4 by 256 and actually I am trying to capture the motion of vehicles from an egocentric perspective (kitti data is taken from the driver's seat), so the vehicles can sometimes appear a bit small if they are reasonably far away but almost every time the vehicles do come closer and are usually big enough.
- One more thing I noticed was that right after training began, the loss dropped from 16 to 4.5 (right after 5 iterations of the first epoch) and it just stayed there up till the 22nd epoch.

**Comment by sarimmehdi** [comment]
I tried changing many hyperparameters like learning rate and weight decay (increased and decreased both of them). Total images in my training set are 5400 and 2160 in the validation set. With a batch size of 16, that is not so many image sequences to train with, especially with num_seq set to 8 and seq_len set to 5 (I also tried with 3).

I noticed that no matter how much I change the hyperparameters, the loss function gets to 3.8 and the top1 accuracy gets to 0.220 within the first 10 epochs and after that, there is little to no change whatsoever. In fact, I waited around for 300 epochs but the loss and top1 accuracy stayed the same right till the end.

I haven't made any changes to your architecture whatsoever (just cloned your repository and ran the code, I had to write a different CustomDataset class of course to load the kitti images). I think it is possible that your self-supervision probably doesn't work on datasets like kitti? I am honestly quite new to all this (started neural networks two months ago), so I can't understand the real reason here. Maybe you have a better idea of what could possibly be going wrong here?

Thanks

**Comment by TengdaHan** [comment]
Hi. Several reasons.
1. The motion you want to encode (a few pixel change) may be too tiny for our DPC. DPC aims to learn high-level representations, like the action class level, and we deliberately avoid learning low-level features like appearance, texture, etc. But self-supervised tracking probably relies more on these lower-level features. 
2. For video object tracking, I recommend you can check out [CorrFlow](https://github.com/zlai0/CorrFlow) and the [paper](https://arxiv.org/abs/1905.00875). That could be more relevant to your task.

**Comment by sarimmehdi** [comment]
Thank you very much for your help. I will definitely give this a look
